File: openai/typescript/src/services/chat/validation.ts
import { InvalidRequestError } from '../../errors/categories.js';
import type { ChatCompletionRequest } from './types.js';

export class ChatCompletionValidator {
  static validate(request: ChatCompletionRequest): void {
    if (!request.model) {
      throw new InvalidRequestError('model is required', { param: 'model' });
    }

    if (!request.messages || request.messages.length === 0) {
      throw new InvalidRequestError('messages array is required and cannot be empty', { param: 'messages' });
    }

    for (let i = 0; i < request.messages.length; i++) {
      const msg = request.messages[i];
      if (!msg.role) {
        throw new InvalidRequestError(`messages[${i}].role is required`, { param: `messages[${i}].role` });
      }
      if (msg.content === undefined && !msg.tool_calls) {
        throw new InvalidRequestError(`messages[${i}].content or tool_calls is required`, { param: `messages[${i}]` });
      }
    }

    if (request.temperature !== undefined && (request.temperature < 0 || request.temperature > 2)) {
      throw new InvalidRequestError('temperature must be between 0 and 2', { param: 'temperature' });
    }

    if (request.top_p !== undefined && (request.top_p < 0 || request.top_p > 1)) {
      throw new InvalidRequestError('top_p must be between 0 and 1', { param: 'top_p' });
    }

    if (request.max_tokens !== undefined && request.max_tokens < 1) {
      throw new InvalidRequestError('max_tokens must be at least 1', { param: 'max_tokens' });
    }
  }
}

File: openai/rust/src/services/chat/validation.rs
use crate::errors::{OpenAIError, OpenAIResult, ValidationError};
use crate::services::chat::ChatCompletionRequest;

pub struct ChatRequestValidator;

impl ChatRequestValidator {
    pub fn validate(request: &ChatCompletionRequest) -> OpenAIResult<()> {
        if request.messages.is_empty() {
            return Err(OpenAIError::Validation(
                ValidationError::MissingRequiredField("messages".to_string()),
            ));
        }

        if request.model.is_empty() {
            return Err(OpenAIError::Validation(
                ValidationError::MissingRequiredField("model".to_string()),
            ));
        }

        if let Some(temperature) = request.temperature {
            if !(0.0..=2.0).contains(&temperature) {
                return Err(OpenAIError::Validation(ValidationError::ValueOutOfRange {
                    field: "temperature".to_string(),
                    min: "0.0".to_string(),
                    max: "2.0".to_string(),
                    value: temperature.to_string(),
                }));
            }
        }

        if let Some(top_p) = request.top_p {
            if !(0.0..=1.0).contains(&top_p) {
                return Err(OpenAIError::Validation(ValidationError::ValueOutOfRange {
                    field: "top_p".to_string(),
                    min: "0.0".to_string(),
                    max: "1.0".to_string(),
                    value: top_p.to_string(),
                }));
            }
        }

        if let Some(presence_penalty) = request.presence_penalty {
            if !(-2.0..=2.0).contains(&presence_penalty) {
                return Err(OpenAIError::Validation(ValidationError::ValueOutOfRange {
                    field: "presence_penalty".to_string(),
                    min: "-2.0".to_string(),
                    max: "2.0".to_string(),
                    value: presence_penalty.to_string(),
                }));
            }
        }

        if let Some(frequency_penalty) = request.frequency_penalty {
            if !(-2.0..=2.0).contains(&frequency_penalty) {
                return Err(OpenAIError::Validation(ValidationError::ValueOutOfRange {
                    field: "frequency_penalty".to_string(),
                    min: "-2.0".to_string(),
                    max: "2.0".to_string(),
                    value: frequency_penalty.to_string(),
                }));
            }
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::services::chat::ChatMessage;

    #[test]
    fn test_validate_valid_request() {
        let request = ChatCompletionRequest::new("gpt-4", vec![ChatMessage::user("Hello")]);
        assert!(ChatRequestValidator::validate(&request).is_ok());
    }

    #[test]
    fn test_validate_empty_messages() {
        let request = ChatCompletionRequest::new("gpt-4", vec![]);
        assert!(ChatRequestValidator::validate(&request).is_err());
    }

    #[test]
    fn test_validate_invalid_temperature() {
        let request = ChatCompletionRequest::new("gpt-4", vec![ChatMessage::user("Hello")])
            .with_temperature(3.0);
        assert!(ChatRequestValidator::validate(&request).is_err());
    }
}
